server/engine/engine.go

```diff
@@ -153,6 +153,23 @@ func (engine *Engine) startDeploymentExecutions() {
 	}
 }
 
+func (engine *Engine) SetFinalStatus() {
+	steps := []model.Step{}
+	engine.database.Instance.Model(&model.Step{}).Preload("Executions").Find(&steps)
+	status := model.Succeeded
+	for _, step := range steps {
+		latestExecution := engine.GetLatestExecution(step)
+		if latestExecution.Status == model.PermanentlyFailed {
+			status = model.Failed
+			break
+		} else if latestExecution.Status == model.Canceled {
+			status = model.Canceled
+			break
+		}
+		model.SetMetric(engine.database.Instance, model.DeployStatus, string(status))
```

Comment:
Ha ha, SetMetric is in the loop:

sqlite> select * from telemetries;
1|2023-04-06 17:55:39.419098323+00:00|2023-04-06 17:55:39.419098323+00:00|0001-01-01 00:00:00+00:00|DeployStatus|Succeeded|
2|2023-04-06 17:55:39.498874701+00:00|2023-04-06 17:55:39.498874701+00:00|0001-01-01 00:00:00+00:00|DeployStatus|Succeeded|
3|2023-04-06 17:55:39.577004524+00:00|2023-04-06 17:55:39.577004524+00:00|0001-01-01 00:00:00+00:00|DeployStatus|Succeeded|
4|2023-04-06 17:55:39.648886137+00:00|2023-04-06 17:55:39.648886137+00:00|0001-01-01 00:00:00+00:00|DeployStatus|Succeeded|
5|2023-04-06 17:55:39.720511242+00:00|2023-04-06 17:55:39.720511242+00:00|0001-01-01 00:00:00+00:00|DeployStatus|Succeeded|
6|2023-04-06 17:55:39.924866303+00:00|2023-04-06 17:55:39.924866303+00:00|0001-01-01 00:00:00+00:00|DeployStatus|Succeeded|
7|2023-04-06 17:55:40.015527446+00:00|2023-04-06 17:55:40.015527446+00:00|0001-01-01 00:00:00+00:00|DeployStatus|Succeeded|
8|2023-04-06 17:55:40.069384455+00:00|2023-04-06 17:55:40.069384455+00:00|0001-01-01 00:00:00+00:00|DeployStatus|Succeeded|
9|2023-04-06 17:55:40.142963425+00:00|2023-04-06 17:55:40.142963425+00:00|0001-01-01 00:00:00+00:00|DeployStatus|Succeeded|
10|2023-04-06 17:55:40.240783409+00:00|2023-04-06 17:55:40.240783409+00:00|0001-01-01 00:00:00+00:00|DeployStatus|Succeeded|
sqlite>

Reply:
removed the setmetric from the loop